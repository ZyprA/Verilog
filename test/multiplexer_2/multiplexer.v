module multiplexer(din, sel, dout);
    input [2:0] din;
    input [1:0] sel;
    output dout;

    function mux;
        input [2:0] din;
        input [1:0] sel;

        if (sel == 2'b00)
            mux = din[0];
        else if (sel == 2'b01)
            mux = din[1];
        else if (sel == 2'b10)
            mux = din[2];
        else
            mux = 1'bx;
    endfunction

    assign dout = mux(din, sel);

endmodule

module multiplexer_sim;
    reg [2:0] din;
    reg [1:0] sel;
    wire dout;

    multiplexer uut(din, sel, dout);

    initial begin
        $dumpfile("multiplexer.vcd");
        $dumpvars(0, multiplexer_sim);
    end

    initial begin
        din[0] = 0; din[1] = 0; din[2] = 1; sel = 0;
        #100 sel = 0;
        #100 sel = 2;
        #100 sel = 1;
        #100 sel = 3;
        #100 $finish;
    end

endmodule