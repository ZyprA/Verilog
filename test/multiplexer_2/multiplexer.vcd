$date
	Sun Feb  2 15:42:56 2025
$end
$version
	Icarus Verilog
$end
$timescale
	1s
$end
$scope module multiplexer_sim $end
$var wire 1 ! dout $end
$var reg 3 " din [2:0] $end
$var reg 2 # sel [1:0] $end
$scope module uut $end
$var wire 3 $ din [2:0] $end
$var wire 2 % sel [1:0] $end
$var wire 1 ! dout $end
$scope function mux $end
$var reg 3 & din [2:0] $end
$var reg 2 ' sel [1:0] $end
$upscope $end
$upscope $end
$upscope $end
$enddefinitions $end
$comment Show the parameter values. $end
$dumpall
$end
#0
$dumpvars
b0 '
b100 &
b0 %
b100 $
b0 #
b100 "
0!
$end
#200
1!
b10 '
b10 #
b10 %
#300
0!
b1 '
b1 #
b1 %
#400
x!
b11 '
b11 #
b11 %
#500
