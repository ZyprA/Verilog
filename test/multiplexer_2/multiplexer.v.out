#! /opt/homebrew/Cellar/icarus-verilog/12.0/bin/vvp
:ivl_version "12.0 (stable)";
:ivl_delay_selection "TYPICAL";
:vpi_time_precision + 0;
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/system.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/vhdl_sys.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/vhdl_textio.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/v2005_math.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/va_math.vpi";
S_0x121609630 .scope module, "multiplexer_sim" "multiplexer_sim" 2 24;
 .timescale 0 0;
v0x12161f500_0 .var "din", 2 0;
v0x12161f5b0_0 .net "dout", 0 0, L_0x12161f730;  1 drivers
v0x12161f660_0 .var "sel", 1 0;
S_0x1216097a0 .scope module, "uut" "multiplexer" 2 29, 2 1 0, S_0x121609630;
 .timescale 0 0;
    .port_info 0 /INPUT 3 "din";
    .port_info 1 /INPUT 2 "sel";
    .port_info 2 /OUTPUT 1 "dout";
v0x12161f290_0 .net "din", 2 0, v0x12161f500_0;  1 drivers
v0x12161f350_0 .net "dout", 0 0, L_0x12161f730;  alias, 1 drivers
v0x12161f3f0_0 .net "sel", 1 0, v0x12161f660_0;  1 drivers
L_0x12161f730 .ufunc/vec4 TD_multiplexer_sim.uut.mux, 1, v0x12161f500_0, v0x12161f660_0 (v0x12160f4e0_0, v0x12161f1d0_0) S_0x121607530;
S_0x121607530 .scope function.vec4.s1, "mux" "mux" 2 6, 2 6 0, S_0x1216097a0;
 .timescale 0 0;
v0x12160f4e0_0 .var "din", 2 0;
; Variable mux is vec4 return value of scope S_0x121607530
v0x12161f1d0_0 .var "sel", 1 0;
TD_multiplexer_sim.uut.mux ;
    %load/vec4 v0x12161f1d0_0;
    %cmpi/e 0, 0, 2;
    %jmp/0xz  T_0.0, 4;
    %load/vec4 v0x12160f4e0_0;
    %parti/s 1, 0, 2;
    %ret/vec4 0, 0, 1;  Assign to mux (store_vec4_to_lval)
    %jmp T_0.1;
T_0.0 ;
    %load/vec4 v0x12161f1d0_0;
    %cmpi/e 1, 0, 2;
    %jmp/0xz  T_0.2, 4;
    %load/vec4 v0x12160f4e0_0;
    %parti/s 1, 1, 2;
    %ret/vec4 0, 0, 1;  Assign to mux (store_vec4_to_lval)
    %jmp T_0.3;
T_0.2 ;
    %load/vec4 v0x12161f1d0_0;
    %cmpi/e 2, 0, 2;
    %jmp/0xz  T_0.4, 4;
    %load/vec4 v0x12160f4e0_0;
    %parti/s 1, 2, 3;
    %ret/vec4 0, 0, 1;  Assign to mux (store_vec4_to_lval)
    %jmp T_0.5;
T_0.4 ;
    %pushi/vec4 1, 1, 1;
    %ret/vec4 0, 0, 1;  Assign to mux (store_vec4_to_lval)
T_0.5 ;
T_0.3 ;
T_0.1 ;
    %end;
    .scope S_0x121609630;
T_1 ;
    %vpi_call 2 32 "$dumpfile", "multiplexer.vcd" {0 0 0};
    %vpi_call 2 33 "$dumpvars", 32'sb00000000000000000000000000000000, S_0x121609630 {0 0 0};
    %end;
    .thread T_1;
    .scope S_0x121609630;
T_2 ;
    %pushi/vec4 0, 0, 1;
    %ix/load 4, 0, 0;
    %flag_set/imm 4, 0;
    %store/vec4 v0x12161f500_0, 4, 1;
    %pushi/vec4 0, 0, 1;
    %ix/load 4, 1, 0;
    %flag_set/imm 4, 0;
    %store/vec4 v0x12161f500_0, 4, 1;
    %pushi/vec4 1, 0, 1;
    %ix/load 4, 2, 0;
    %flag_set/imm 4, 0;
    %store/vec4 v0x12161f500_0, 4, 1;
    %pushi/vec4 0, 0, 2;
    %store/vec4 v0x12161f660_0, 0, 2;
    %delay 100, 0;
    %pushi/vec4 0, 0, 2;
    %store/vec4 v0x12161f660_0, 0, 2;
    %delay 100, 0;
    %pushi/vec4 2, 0, 2;
    %store/vec4 v0x12161f660_0, 0, 2;
    %delay 100, 0;
    %pushi/vec4 1, 0, 2;
    %store/vec4 v0x12161f660_0, 0, 2;
    %delay 100, 0;
    %pushi/vec4 3, 0, 2;
    %store/vec4 v0x12161f660_0, 0, 2;
    %delay 100, 0;
    %vpi_call 2 42 "$finish" {0 0 0};
    %end;
    .thread T_2;
# The file index is used to find the file name in the following table.
:file_names 3;
    "N/A";
    "<interactive>";
    "multiplexer.v";
